### PR TITLE
CAL-104 configure DefaultSecurityAttributeValuesPlugin to apply secur…

### PR DESCRIPTION
**Pull request by ryeats**
#### What does this PR do?
 if no security attributes are present this plugin adds security attributes to metacards based on the system users attributes.
#### Who is reviewing it?
@bcwaters @coyotesqrl @tbatie @adimka @bdeining @jaymcnallie  @stustison 
#### How should this be tested
deploy a standard profile alliance build
#### What are the relevant tickets?
DDF-2291
#### Checklist:
- [ ] Documentation Updated
- [ x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

…ity defaults to metacards based on the system high-water mark

### CODE REVIEW

File: catalog/security/security-app/src/main/resources/features.xml [hunk 1]
```diff
@@ -21,9 +21,16 @@
...
+        <feature>catalog-plugin-defaultsecurityattributevalues</feature>
+        <configfile finalname="/etc/ddf.catalog.metacard.security.defaultsecurityattributevaluesplugin.config" override="true">
+            mvn:org.codice.alliance.security/security-app/${project.version}/config/defaultsecurityattributevalues
+        </configfile>
+    </feature>
+
     <feature name="security-app" install="auto" version="${project.version}"
              description="The Alliance Security App provides features to enhance security. ::Alliance Security">
         <feature prerequisite="true">catalog-app</feature>
-        <feature>banner-marking</feature>
```
Comment (bdeining) [inline catalog/security/security-app/src/main/resources/features.xml]: Just curious : Why was this removed?

**Comment by bdeining** [comment]
Is this CAL-104?

File: catalog/security/security-app/src/main/resources/features.xml (hunk 1 above)
Comment (coyotesqrl) [inline catalog/security/security-app/src/main/resources/features.xml]: I think the `banner-marking` feature should be marked `install="auto"` with a prereq on `security-app` so it is always on when `security-app` is installed. Thoughts @jlcsmith?

Comment (jlcsmith) [inline catalog/security/security-app/src/main/resources/features.xml]: That makes sense to me

**Comment by glenhein** [comment]
I cleared my local mvn repo of org/codice/alliance jars, and then ran ```mvn clean verify```. I got the error

```
[ERROR] Failed to execute goal org.apache.karaf.tooling:karaf-maven-plugin:4.0.4:kar (kar) on project security-app: Failed to create archive: Could not find artifact org.codice.alliance.security:security-app:config:defaultsecurityattributevalues:0.2-SNAPSHOT in office (http://nexus:8081/nexus/content/groups/public/)
[ERROR] 
[ERROR] Try downloading the file manually from the project website.
[ERROR] 
[ERROR] Then, install it using the command:
[ERROR] mvn install:install-file -DgroupId=org.codice.alliance.security -DartifactId=security-app -Dversion=0.2-SNAPSHOT -Dclassifier=defaultsecurityattributevalues -Dpackaging=config -Dfile=/path/to/file
[ERROR] 
[ERROR] Alternatively, if you host your own repository you can deploy the file there:
[ERROR] mvn deploy:deploy-file -DgroupId=org.codice.alliance.security -DartifactId=security-app -Dversion=0.2-SNAPSHOT -Dclassifier=defaultsecurityattributevalues -Dpackaging=config -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
```

**Comment by ryeats** [comment]
Thanks glen for the heads up fixed that the plugins were in the wrong order. I have to move alot of code into this PR turns out I need the whole plugin to live in Alliance instead of ddf so hopefully Ill have that up today.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.codice.alliance.catalog.core</groupId>
+            <artifactId>catalog-core-api-impl</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <!--<dependency>-->
+            <!--<groupId>org.osgi</groupId>-->
+            <!--<artifactId>org.osgi.core</artifactId>-->
+        <!--</dependency>-->
```
Comment (bdeining) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml]: These commented dependencies can be removed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.security.assertion.SecurityAssertion;
+
+public class DefaultSecurityAttributeValuesPlugin implements PreIngestPlugin {
+
+    private static final SecurityAttributes SECURITY_ATTRIBUTE_DESCRIPTORS =
+            new SecurityAttributes();
+
+    //// TODO (RWY) - is this tag name good
+    public static final String DEFAULTMARKINGS = "defaultmarkings";
```
Comment (bdeining) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: Recommend using camelCase here

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/resources/OSGI-INF/metatype/metatype.xml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+			required="true"
+			type="String" 
+			default=""
+			description="Mappings of system attributes to metacard markings.">
+		</AD>
+		<AD name="Owner Producer" id="ownerProducer"
+			required="true"
+			type="String" 
+			default="DutyOrganization"
+			description="Mappings of system attributes to metacard markings.">
```
Comment (bdeining) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/resources/OSGI-INF/metatype/metatype.xml]: Looks like these descriptions need to be updated

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        when(mockedAssertion.getAttributeStatements()).thenReturn(Arrays.asList(
+                mockedAttributeStatement1,
+                mockedAttributeStatement2));
+        when(mockedAttributeStatement1.getAttributes()).thenReturn(Arrays.asList(mockedAttribute1));
+        when(mockedAttributeStatement2.getAttributes()).thenReturn(Arrays.asList(mockedAttribute2));
+        when(mockedXss1.getValue()).thenReturn("1");
+        when(mockedXss2.getValue()).thenReturn("2");
+        when(mockedAttribute1.getAttributeValues()).thenReturn(Arrays.asList(mockedXss1));
+        when(mockedAttribute2.getAttributeValues()).thenReturn(Arrays.asList(mockedXss2));
+        when(mockedAttribute1.getName()).thenReturn("UserAttrib1");
```
Comment (bdeining) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java]: `UserAttrib1` and `UserAttrib2` can be made into constants

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    @Test
+    public void testProcessCreateRequest() throws Exception {
+        CreateRequest modifiedRequest = defaultSecurityAttributeValuesPlugin.process(createRequest);
+        assertThat(createRequest, not(modifiedRequest));
+    }
+
+    @Test
+    public void testProcessUpdateREquest() throws Exception {
```
Comment (bdeining) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java]: `testProcessUpdateRequest`

**Comment by glenhein** [comment]
I cleared my local mvn repo of org/codice/alliance jars, and then ran mvn clean install. Everything built as expected. I was able to run Alliance and ingest test content.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml (hunk 2 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml]: fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java (hunk 3 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java (hunk 5 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java]: fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java (hunk 6 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java]: fixed

File: catalog/security/security-app/src/main/resources/features.xml (hunk 1 above)
Comment (ryeats) [inline catalog/security/security-app/src/main/resources/features.xml]: fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+                unmarkedMetacard)));
+        when(emptySecurityAttributeList.getValue()).thenReturn(mockedEmptyAttributeMap);
+        when(mockedEmptyAttributeMap.isEmpty()).thenReturn(true);
+    }
+
+    @Test
+    public void testProcessUnmarkedMetacard() throws Exception {
+        Metacard modifiedMetacard = defaultSecurityAttributeValuesPlugin.addDefaults(
+                unmarkedMetacard);
+        assertThat(unmarkedMetacard, not(modifiedMetacard));
```
Comment (jlcsmith) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java]: why not validate the contents of the modified metacard match the expected values?

File: catalog/security/security-app/src/main/resources/etc/org.codice.ddf.catalog.security.policy.metacard.MetacardAttributeSecurityPolicyPlugin.config [hunk 8]
```diff
@@ -0,0 +1,11 @@
+intersectMetacardAttributes=[ \
+  ]
+service.pid="org.codice.ddf.catalog.security.policy.metacard.MetacardAttributeSecurityPolicyPlugin"
+unionMetacardAttributes=[ \
+  "security.classification\=classification", \
+  "security.codewords\=SCIcontrols", \
+  "security.dissemination-controls\=disseminationControls", \
+  "security.security.other-dissemination-controls\=nonICmarkings", \
+  "security.owner-producer\=releasableTo", \
```
Comment (jlcsmith) [inline catalog/security/security-app/src/main/resources/etc/org.codice.ddf.catalog.security.policy.metacard.MetacardAttributeSecurityPolicyPlugin.config]: This doesn't look like the correct mapping

File: catalog/security/security-app/src/main/resources/etc/org.codice.ddf.catalog.security.policy.metacard.MetacardAttributeSecurityPolicyPlugin.config [hunk 9]
```diff
@@ -0,0 +1,11 @@
+intersectMetacardAttributes=[ \
+  ]
+service.pid="org.codice.ddf.catalog.security.policy.metacard.MetacardAttributeSecurityPolicyPlugin"
+unionMetacardAttributes=[ \
+  "security.classification\=classification", \
+  "security.codewords\=SCIcontrols", \
```
Comment (jlcsmith) [inline catalog/security/security-app/src/main/resources/etc/org.codice.ddf.catalog.security.policy.metacard.MetacardAttributeSecurityPolicyPlugin.config]: should this be SCIcontrols or FineAccessControls?

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+                .stream()
+                .filter(Objects::nonNull)
+                .forEach(m -> updatedMetacards.add(addDefaults(m)));
+        return new CreateRequestImpl(updatedMetacards,
+                createRequest.getProperties(),
+                createRequest.getStoreIds());
+    }
+
+    @Override
+    public UpdateRequest process(UpdateRequest updateRequest)
```
Comment (jlcsmith) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: Shouldn't this plugin process updates as well?  If security attributes are missing, defaults should apply.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java (hunk 7 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/test/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPluginTest.java]: no good reason I was originally only working with mocked objects and ripped out the verify calls and just forgot to add assertions when I changed it. fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java (hunk 10 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: I'm on the fence about this.  I had that originally but updating existing data without markings would cause it to seemingly disappear. That being said if someone exported all of their metacards into a new instance the new instance would mark some of the data differently so their is also something to be said for consistency.

File: catalog/security/security-app/src/main/resources/etc/org.codice.ddf.catalog.security.policy.metacard.MetacardAttributeSecurityPolicyPlugin.config (hunk 8 above)
Comment (ryeats) [inline catalog/security/security-app/src/main/resources/etc/org.codice.ddf.catalog.security.policy.metacard.MetacardAttributeSecurityPolicyPlugin.config]: you are correct, I think the answer is I shouldn't be configuring this here, the mappings for attributes is utterly confusing but we don't have a default AuthzRealm policy config or any  claims configured for sts in alliance so this doesn't make sense.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            <artifactId>commons-lang3</artifactId>
+            <version>${commons-lang3.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.servicemix.bundles</groupId>
+            <artifactId>org.apache.servicemix.bundles.opensaml</artifactId>
+            <version>3.1.1_3</version>
+        </dependency>
+        <dependency>
+            <groupId>com.diffplug.guava</groupId>
```
Comment (coyotesqrl) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml]: What's this?

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Embed-Dependency>
+                            catalog-core-api-impl,
```
Comment (coyotesqrl) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml]: So here's my question and concern with Alliance and DDF module names being the same:

Are you embedding the Alliance library, the DDF library, or both here? What does the manifest show?

Is there a syntax by which you can specify coordinates more fully to the `Embed-Dependency` directive?

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml (hunk 11 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml]: Its really suspicious ill point to the real guava distribution instead

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml (hunk 12 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml]: yes this threw me for a loop and i thought it was causing me problems but it works I embed both,the manifest shows both and the only way not to get both was to not depend on both so if one was a bundle and the other was a jar you would be screwed.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+import ddf.security.assertion.SecurityAssertion;
+
+public class DefaultSecurityAttributeValuesPlugin implements PreIngestPlugin {
+
+    private static final SecurityAttributes SECURITY_ATTRIBUTE_DESCRIPTORS =
+            new SecurityAttributes();
+
+    public static final String DEFAULTMARKINGS = "defaultMarkings";
+
+    private static Multimap<String, String> sysHighToMetacardAttributeMapping =
```
Comment (coyotesqrl) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: I like `Multimap`, but when you work with the values here, you work with them as a Collection; I don't see any value in using a `Multimap` and not a `Map<String, Set<String>>` instead.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        if (!SECURITY_ATTRIBUTE_DESCRIPTORS.getAttributeDescriptors()
+                .stream()
+                .anyMatch(ad -> metacard.getMetacardType()
+                        .getAttributeDescriptors()
+                        .contains(ad))) {
+            MetacardImpl metacardImpl = new MetacardImpl(metacard,
+                    new MetacardTypeImpl(metacard.getMetacardType()
+                            .getName(),
+                            metacard.getMetacardType(),
+                            SECURITY_ATTRIBUTE_DESCRIPTORS.getAttributeDescriptors()));
```
Comment (bdeining) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: You can pass a list of `MetacardType`s into the `MetacardTypeImpl` constructor instead of having to deal with a set of `AttributeDescriptors`.  https://github.com/codice/ddf/blob/master/catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardTypeImpl.java#L120.  You may also want to consider creating the `MetacardType` via blueprint

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+     * attribute.
+     */
+    private Map<String, Attribute> getHighwaterSecurityMarkings() {
+        Map<String, Attribute> securityMarkings = new HashMap<>();
+        ddf.security.Subject system =
+                org.codice.ddf.security.common.Security.runAsAdmin(() -> getSystemSubject());
+        SecurityAssertion assertion = system.getPrincipals()
+                .oneByType(SecurityAssertion.class);
+        List<AttributeStatement> attributeStatements = assertion.getAttributeStatements();
+        for (AttributeStatement curStatement : attributeStatements) {
```
Comment (coyotesqrl) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: You can reduce cyclomatic complexity a little bit and improve overall performance with a couple of small tweaks. Take a look at this version of this block:
```java
        for (AttributeStatement curStatement : attributeStatements) {
            for (org.opensaml.saml.saml2.core.Attribute attribute : curStatement.getAttributes()) {
                Collection<String> attributeNames =
                        sysHighToMetacardAttributeMapping.get(attribute.getName());

                if (attributeNames.isEmpty()) {
                    continue;
                }
                HashSet<Serializable> attributeValues = attribute.getAttributeValues()
                        .stream()
                        .filter(curValue -> curValue instanceof XSString)
                        .map(XSString.class::cast)
                        .map(XSString::getValue)
                        .collect(Collectors.toCollection(HashSet::new));

                for (String attributeName : attributeNames) {
                    HashSet<Serializable> values = new HashSet<>(attributeValues);

                    securityMarkings.compute(attributeName, (k, v) -> {
                        if (v != null) {
                            values.addAll(v.getValues());
                        }
                        return new AttributeImpl(k,
                                (List<Serializable>) ImmutableList.copyOf(values));
                    });
                }
            }
        }
```

Comment (coyotesqrl) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: The biggest improvement is moving the streaming creation of the `attributeValues` set out of the for loop and then just getting a new copy when looping. The inversion of the test on `attributeNames.isEmpty()` reduces complexity. 

The `compute` is there because I was hoping I could find a way to make it more efficient but it's basically a wash.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+                                                values)));
+                    }
+                }
+            }
+        }
+        return securityMarkings;
+    }
+
+    /**
+     * adds default system high-water markings in the event that none of the policies were able to apply security markings to this metacard
```
Comment (coyotesqrl) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: Capitalize; add period.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+                            .getName(),
+                            metacard.getMetacardType(),
+                            SECURITY_ATTRIBUTE_DESCRIPTORS.getAttributeDescriptors()));
+            Set<String> updatedTags = new HashSet<>(metacard.getTags());
+            updatedTags.add(DEFAULTMARKINGS);
+            metacardImpl.setTags(updatedTags);
+            extendedMetacard = metacardImpl;
+        } else {
+            MetacardImpl metacardImpl = new MetacardImpl(metacard);
+            Set<String> updatedTags = new HashSet<>(metacard.getTags());
```
Comment (coyotesqrl) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: These four lines are identical to 133-146; they can be pulled out of the if/else.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    public void setOwnerProducer(String ownerProducer) {
+        sysHighToMetacardAttributeMapping.put(ownerProducer, Security.OWNER_PRODUCER);
+    }
+
+    @Override
+    public CreateRequest process(CreateRequest createRequest)
+            throws PluginExecutionException, StopProcessingException {
+        List<Metacard> updatedMetacards = new ArrayList<>();
```
Comment (coyotesqrl) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]:
```java
        List<Metacard> updatedMetacards = createRequest.getMetacards()
                .stream()
                .filter(Objects::nonNull)
                .map(this::addDefaults)
                .collect(Collectors.toList());
```

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java (hunk 14 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: I created the MetacardType via blueprint. Its actually easier to pass the AttributeDescriptors maybe we should add a new constructor to MetacardTypeImpl(String name, MetacardType... metacardTypes) however I can't do it in this PR since i don't have ddf changes.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java (hunk 16 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java (hunk 13 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java (hunk 15 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: added the attributeNames.isEmpty() change.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java (hunk 17 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java (hunk 18 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java]: fixed

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/resources/OSGI-INF/metatype/metatype.xml (hunk 4 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/resources/OSGI-INF/metatype/metatype.xml]: fixed

**Comment by coyotesqrl** [comment]
👍

**Comment by tbatie** [comment]
:+1:

**Comment by adimka** [comment]
👍

**Comment by stustison** [comment]
👍

**Comment by adimka** [comment]
Hero'd successfully.

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml (hunk 12 above)
Comment (jlcsmith) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml]: would it be preferable to explicitly call out both?  Eg: https://github.com/codice/alliance/blob/master/catalog/core/catalog-core-metacardtypes/pom.xml#L61

**Comment by jlcsmith** [comment]
@ryeats LGTM 👍

File: catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml (hunk 12 above)
Comment (ryeats) [inline catalog/plugin/catalog-plugin-defaultsecurityattributevalues/pom.xml]: fixed
